### ISSUE TITLE
IBM powerpc compile error

### ISSUE THREAD

**Issue by Mon-ius** [InsightSoftwareConsortium/ITK#1960]
Hi, I am working on Power9 with ppcle-64 arch. I tried to compile it. It shows error.

```
Scanning dependencies of target ITKReview-all
[100%] Built target ITKReview-all
[100%] Linking CXX static library ../../../../lib/libITKTestKernel-5.1.a
[100%] Built target ITKTestKernel
Scanning dependencies of target itkTestDriver
[100%] Building CXX object Modules/Core/TestKernel/src/CMakeFiles/itkTestDriver.dir/itkTestDriver.cxx.o
[100%] Linking CXX executable ../../../../bin/itkTestDriver
../../../../lib/libitkpng-5.1.a(pngrutil.c.o):pngrutil.c:function itk_png_read_filter_row: error: undefined reference to 'png_init_filter_functions_vsx'
collect2: error: ld returned 1 exit status
make[5]: *** [bin/itkTestDriver] Error 1
make[4]: *** [Modules/Core/TestKernel/src/CMakeFiles/itkTestDriver.dir/all] Error 2
make[3]: *** [all] Error 2
make[2]: *** [ITKv5-prefix/src/ITKv5-stamp/ITKv5-build] Error 2
make[1]: *** [CMakeFiles/ITKv5.dir/all] Error 2
make: *** [all] Error 2
```

**Comment by dzenanz** [comment]
Does [updating the bundled PNG](https://github.com/InsightSoftwareConsortium/ITK/pull/1962) version help? I suspect that [platform-specific optimizations](https://github.com/InsightSoftwareConsortium/ITK/blob/4bdc38ca9bedac7c2c61e29f014e902e1fe6a2e7/Modules/ThirdParty/PNG/src/itkpng/pngrutil.c#L4130-L4145) in libPNG are the cause.

**Comment by Mon-ius** [comment]
How to Using the `UpdateFromUpstream.sh` script? I didn't find it.

**Comment by dzenanz** [comment]
https://github.com/InsightSoftwareConsortium/ITK/blob/master/Modules/ThirdParty/PNG/UpdateFromUpstream.sh

**Comment by Mon-ius** [comment]
I tried to replace the file from https://raw.githubusercontent.com/InsightSoftwareConsortium/ITK/4bdc38ca9bedac7c2c61e29f014e902e1fe6a2e7/Modules/ThirdParty/PNG/src/itkpng/pngrutil.c, and still error.

`../../../../lib/libitkpng-5.2.a(pngrutil.c.o):pngrutil.c:function itk_png_read_filter_row: error: undefined reference to 'png_init_filter_functions_vsx'`

How to use your script? `UpdateFromUpstream.sh` after `git clone https://github.com/InsightSoftwareConsortium/ITK.git`?

**Comment by Mon-ius** [comment]
I tried to run the `UpdateFromUpstream.sh` under the project root folder.

`./Modules/ThirdParty/PNG/UpdateFromUpstream.sh`

And get these.
```

fatal: cannot use Perl-compatible regexes when not compiled with USE_LIBPCRE
Created upstream snapshot branch 'upstream-png'.
error: Your local changes to the following files would be overwritten by merge:
	Modules/ThirdParty/PNG/src/itkpng/LICENSE
	Modules/ThirdParty/PNG/src/itkpng/png.c
	Modules/ThirdParty/PNG/src/itkpng/png.h
	Modules/ThirdParty/PNG/src/itkpng/pngconf.h
	Modules/ThirdParty/PNG/src/itkpng/pngdebug.h
	Modules/ThirdParty/PNG/src/itkpng/pngerror.c
	Modules/ThirdParty/PNG/src/itkpng/pngget.c
	Modules/ThirdParty/PNG/src/itkpng/pnginfo.h
	Modules/ThirdParty/PNG/src/itkpng/pngmem.c
	Modules/ThirdParty/PNG/src/itkpng/pngpread.c
	Modules/ThirdParty/PNG/src/itkpng/pngpriv.h
	Modules/ThirdParty/PNG/src/itkpng/pngread.c
	Modules/ThirdParty/PNG/src/itkpng/pngrio.c
	Modules/ThirdParty/PNG/src/itkpng/pngrtran.c
	Modules/ThirdParty/PNG/src/itkpng/pngrutil.c
	Modules/ThirdParty/PNG/src/itkpng/pngset.c
	Modules/ThirdParty/PNG/src/itkpng/pngstruct.h
	Modules/ThirdParty/PNG/src/itkpng/pngtest.c
	Modules/ThirdParty/PNG/src/itkpng/pngtrans.c
	Modules/ThirdParty/PNG/src/itkpng/pngwio.c
	Modules/ThirdParty/PNG/src/itkpng/pngwrite.c
	Modules/ThirdParty/PNG/src/itkpng/pngwtran.c
	Modules/ThirdParty/PNG/src/itkpng/pngwutil.c
Please commit your changes or stash them before you merge.
Aborting
error: The branch 'upstream-png' is not fully merged.
If you are sure you want to delete it, run 'git branch -D upstream-png'.
```

**Comment by dzenanz** [comment]
In #1962 I did a git clone, then `cd Modules/ThirdParty/PNG` and finally `./UpdateFromUpstream.sh`. This resulted in a merge conflict over LICENSE file, which I resolved manually by choosing the existing file.

**Comment by Mon-ius** [comment]
I saw there is an update for https://github.com/InsightSoftwareConsortium/ITK/pull/1962. Does that mean it has no conflict now?

**Comment by Mon-ius** [comment]
Let conclude that:
-  `cd Modules/ThirdParty/PNG`
- `rm src/itkpng/LICENSE `
- `./UpdateFromUpstream.sh`
- `cd $project_root`
-  `mkdir build && cd build && cmake .. && make`

right?

**Comment by dzenanz** [comment]
I manually resolved the merge conflict in #1962. But there were not a lot of changes in this update, so I doubt it will resolve your issue.

**Comment by dzenanz** [comment]
-  `cd Modules/ThirdParty/PNG`
- `./UpdateFromUpstream.sh`
- manually resolve the conflict (I use TortoiseGit so I don't know command line command for this)
- `cd $project_root`
-  `mkdir build && cd build && cmake .. && make`

... or you could git pull my branch ([updatePNG](https://github.com/dzenanz/ITK/tree/updatePNG)) and don't deal with updating.

**Comment by Mon-ius** [comment]
So, `git clone https://github.com/dzenanz/ITK/tree/updatePNG`. And `git checkout updatePNG`, right?

**Comment by dzenanz** [comment]
The checkout steps seem fine. But that does not resolve your error. `png_init_filter_functions_vsx` does not exist on my computer. It seems to come from here:
https://github.com/InsightSoftwareConsortium/ITK/blob/4bdc38ca9bedac7c2c61e29f014e902e1fe6a2e7/Modules/ThirdParty/PNG/src/itkpng/pngpriv.h#L268-L271

As far as I can tell, this is an issue in libPNG. Can you try enabling `ITK_USE_SYSTEM_PNG`? To do that, replace `cmake ..` by `cmake -DITK_USE_SYSTEM_PNG=ON ..` in your build command.

**Comment by Mon-ius** [comment]
Yes, I tried that before raised the issues. And now, I tried again with your repo. No current error, but a new one.

```
[ 65%] Built target ITKImageFilterBaseTestDriver
[ 65%] Building CXX object Modules/Core/Common/test/CMakeFiles/ITKCommon2TestDriver.dir/itkThreadedIteratorRangePartitionerTest2.cxx.o
[ 65%] Building CXX object Modules/Core/Transform/test/CMakeFiles/ITKTransformTestDriver.dir/itkEuler3DTransformTest.cxx.o
[ 65%] Building CXX object Modules/Core/ImageFunction/test/CMakeFiles/ITKImageFunctionTestDriver.dir/itkGaussianDerivativeImageFunctionTest.cxx.o
[ 65%] Building CXX object Modules/IO/ImageBase/test/CMakeFiles/ITKIOImageBaseTestDriver.dir/itkImageFileWriterTest.cxx.o
/castles/nr/projects/d/duanj-ai-in-medical-imaging/ITK/Modules/IO/BMP/test/itkBMPImageIOTestExtension.cxx: In function ‘bool {anonymous}::WriteSmallBmp(const string&)’:
/castles/nr/projects/d/duanj-ai-in-medical-imaging/ITK/Modules/IO/BMP/test/itkBMPImageIOTestExtension.cxx:33:78: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ inside { } [-Wnarrowing]
                            0,  0,  0,  1, 0, 1, 0, 1, 0, 24, 0,  0, 0, -1, 0 };
                                                                              ^
[ 65%] Linking CXX executable ../../../../bin/ITKAnisotropicSmoothingTestDriver
make[2]: *** [Modules/IO/BMP/test/CMakeFiles/ITKIOBMPTestDriver.dir/itkBMPImageIOTestExtension.cxx.o] Error 1
make[2]: *** Waiting for unfinished jobs....
```

```
[ 65%] Linking CXX executable ../../../../bin/ITKAnisotropicSmoothingTestDriver
make[2]: *** [Modules/IO/BMP/test/CMakeFiles/ITKIOBMPTestDriver.dir/itkBMPImageIOTestExtension.cxx.o] Error 1
make[2]: *** Waiting for unfinished jobs....
[ 65%] Building CXX object Modules/IO/GDCM/test/CMakeFiles/ITKIOGDCMTestDriver.dir/itkGDCMSeriesReadImageWriteTest.cxx.o
[ 65%] Built target ITKAnisotropicSmoothingTestDriver
Scanning dependencies of target ITKQuadEdgeMeshTestDriver
[ 65%] Building CXX object Modules/Core/ImageFunction/test/CMakeFiles/ITKImageFunctionTestDriver.dir/itkCentralDifferenceImageFunctionTest.cxx.o
[ 65%] Building CXX object Modules/Numerics/Statistics/test/CMakeFiles/ITKStatisticsTestDriver.dir/ITKStatisticsTestDriver.cxx.o
[ 65%] Linking CXX executable ../../../../bin/ITKImageGridGTestDriver
[ 65%] Built target ITKImageGridGTestDriver
[ 65%] Building CXX object Modules/Numerics/Statistics/test/CMakeFiles/ITKStatisticsTestDriver.dir/itkDecisionRuleTest.cxx.o
[ 66%] Building CXX object Modules/Numerics/Statistics/test/CMakeFiles/ITKStatisticsTestDriver.dir/itkDenseFrequencyContainer2Test.cxx.o
[ 66%] Building CXX object Modules/IO/GDCM/test/CMakeFiles/ITKIOGDCMTestDriver.dir/itkGDCMSeriesMissingDicomTagTest.cxx.o
[ 66%] Building CXX object Modules/Core/Common/test/CMakeFiles/ITKCommon1TestDriver.dir/ITKCommon1TestDriver.cxx.o
[ 66%] Building CXX object Modules/IO/GDCM/test/CMakeFiles/ITKIOGDCMTestDriver.dir/itkGDCMSeriesStreamReadImageWriteTest.cxx.o
[ 66%] Building CXX object Modules/IO/GDCM/test/CMakeFiles/ITKIOGDCMTestDriver.dir/itkGDCMImagePositionPatientTest.cxx.o
[ 67%] Building CXX object Modules/Core/Common/test/CMakeFiles/ITKCommon1TestDriver.dir/itkImageRegionExplicitTest.cxx.o
[ 67%] Building CXX object Modules/Core/Common/test/CMakeFiles/ITKCommon2TestDriver.dir/itkThreadedIteratorRangePartitionerTest3.cxx.o
[ 67%] Building CXX object Modules/Core/Common/test/CMakeFiles/ITKCommonGTestDriver.dir/itkImageBaseGTest.cxx.o
make[1]: *** [Modules/IO/BMP/test/CMakeFiles/ITKIOBMPTestDriver.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```

```
[ 74%] Linking CXX executable ../../../../bin/ITKStatisticsTestDriver
[ 74%] Built target ITKStatisticsTestDriver
[ 74%] Linking CXX executable ../../../../bin/ITKCommon1TestDriver
[ 74%] Built target ITKCommon1TestDriver
make: *** [all] Error 2
```

**Comment by dzenanz** [comment]
I guess a bit more thought needs to be spent about the error and its fix in a multi-platform manner.

**Comment by dzenanz** [comment]
Fixed by 9b7a28032ef46034f91cdc98bb99f4b3f24f270c.
